upstream/upstream.go

```diff
@@ -655,10 +667,102 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 	}
 
 	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	jrReq, err := req.JsonRpcRequest()
+	if err != nil {
+		return fmt.Errorf("failed to get json-rpc request: %w", err), false
+	}
+
+	// sample of methods with block param
+	methodsWithBlockParam := map[string]int{
+		"eth_getBalance":          1,
+		"eth_getCode":             1,
+		"eth_getTransactionCount": 1,
+		"eth_call":                1,
+	}
+
+	if index, exists := methodsWithBlockParam[jrReq.Method]; exists {
+		if len(jrReq.Params) > index {
+			// TODO: do we need to check if blockNumber is a valid hex?
+			_, ok := jrReq.Params[index].(string)
+
+			if ok {
+				nodeType := u.config.Evm.NodeType
+				if nodeType == "" {
+					nodeType, err = u.detectNodeType()
+					if err != nil {
+						return fmt.Errorf("failed to detect node type: %w", err), false
+					}
+				}
+
+				if nodeType == common.EvmNodeTypeFull {
+					lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+					bn, ebn := req.EvmBlockNumber()
+					if ebn != nil {
+						return fmt.Errorf("failed to get request block number: %w", ebn), false
+					}
+
+					if bn < lb-int64(u.config.Evm.MaxAvailableRecentBlocks) {
+						return fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-int64(u.config.Evm.MaxAvailableRecentBlocks)), true
+					}
+				}
+			}
+		}
+	}
 
 	return nil, false
 }
 
+func isMissingHistoricalStateError(err error) bool {
+	if err == nil {
+		return false
+	}
+	errMsg := err.Error()
+	// Check for common error messages indicating missing historical state
+	return strings.Contains(errMsg, "missing trie node") ||
+		strings.Contains(errMsg, "required historical state unavailable") ||
+		strings.Contains(errMsg, "historical state unavailable") ||
+		strings.Contains(errMsg, "state unavailable") ||
+		strings.Contains(errMsg, "not found")
```

Comment:
We have a list in error normalization logic, probably it's best to extract these in common/evm.go and have common.EvmIsMissingDataError so we can use in both places